src/httpserver/http_request.hpp

```diff
@@ -159,29 +159,29 @@ class http_request {
       * @param key the specific header to get the value from
       * @return the value of the header.
      **/
-     const std::string get_header(const std::string& key) const;
+     std::string_view get_header(std::string_view key) const;
```

Comment:
Should we align the rest of the interface of this class to use string_view? (e.g. get_user, get_pass, get_method, etc.)

Reply:
I think that would be best. Design decisions to make though: some of these other methods modify data extracted from MHD (e.g. get_querystring), construct a result and then return it. I see two options here. One is to change the signature and/or return values to match whatever MHD gives, and return views over that data. The other is to construct the result as now only once, cache in the request object, and return a view over that (also if the user calls twice, the cached value is simply returned on the second call). Thoughts?

Reply:
It's unfortunate that I am not sure I really like any of the five options I see for this. We basically can:

1. Leave the logic as is, and have an inconsistent interface. Some methods will return string, others return string_view. The issue with it is that basically we are basically projecting outside our own internal logic and the user pays the cost for this by having to adapt to what each method returns for a reason that is not immediately obvious when in their position.

2. Revert the methods to pure proxies over what libmicrohttpd returns. This comes at a loss of the extra simplicity that httpserver provides (which is one of the core reasons people use this library).

3. Pre-fetch all data from libmicrohttpd and store it preventively in the request object. This would allow to use string_view everywhere. This used to be the old implementation, but we moved away from it because it was just way too expensive for no reason (the user pays the cost of the fetch + copy even when they never use the data).

4. Do a fetch and cache. This is an implementation I had thought of when migrating off of the old prefetch approach. It has the extra advantage of making subsequent calls cheaper. It has a massive issue that it makes all getters non const which is a really annoying property for the interface of what basically is a class meant to store and transport data. I considered and discarded this approach when moving away from the old pre-fetch one.

5. Just leave the class as is and not use string_view at all.

I am currently pondering 1 an 5 - everything else just doesn't seem worth it. Instinctively, I'd say 5 would be my general inclination because philosophically I'd rather prioritize clarity/consistency over marginal performance, but I can welcome disagreements, if well motivated. I also think it would be worth to simulate some heavy data fetch (e.g. of headers) with both approaches to actually prove that using string_view whenever possible is worth the inconsistency it costs interface-wise.

Reply:
I agree that 3 is untenable performance-wise. Most users are not going to use every field, prefetch has a high init cost. 1 is also undesirable; I agree it is better to be as consistent as possible.

I am not sure I agree with your assessment of point 4. An object internal cache doesn't mean we have to sacrifice `const` on all the accessors. It just means we can't cache the items directly in the request object. We can use an private `cache` type, and store it within a `unique_ptr`. Then we have a mutable cache from within a `const` accessor. There are only a few values that need to be cached in this way for us to make the whole API return `string_view`. 

I think I prefer this option, it seems to be the best compromise between API consistency, const-ness, and performance. I have already tried this out, I will add it to the PR and look forward to your thoughts.

Reply:
That's a good idea - I am looking at the implementation and it seems like it lands in a good place. I've just left a few minor comments. Take a look at those. I'll take another re-read later to give you a more complete feedback.